Improve worksheet item manipulation commands

kashizui opened the pull request:
- Remove `cl wadd`
- Add only text items or subworksheets through `cl add`
- Create requirements file for running unit tests
- Fix test-cli issues

Fixes #201

@percyliang @pbhatnagar3

kashizui committed:
Improve worksheet item manipulation commands

 - Remove `cl wadd`
 - Add only text items or subworksheets through `cl add`
 - Create requirements file for running unit tests
 - Fix test-cli issues

Fixes #201

kashizui commented:
cl help add

```
usage: cl add [-h] [-m MESSAGE [MESSAGE ...]] [-w WORKSHEET_SPEC]
              [subworksheet_spec [subworksheet_spec ...]]

Append worksheet links and/or text items to a worksheet.

positional arguments:
  subworksheet_spec     worksheets to add links to
                        ([<alias>::|<address>::]|(<uuid>|<name>))

optional arguments:
  -h, --help            show this help message and exit
  -m MESSAGE [MESSAGE ...], --message MESSAGE [MESSAGE ...]
                        add a text element (must appear after
                        subworksheet_specs)
  -w WORKSHEET_SPEC, --worksheet_spec WORKSHEET_SPEC
                        operate on this worksheet
                        ([<alias>::|<address>::]|(<uuid>|<name>))
```

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -745,11 +744,11 @@ def do_download_command(self, args):
     @Commands.command(
         'cp',
-        help='Copy bundles across instances.',
+        help='Copy bundles across worksheets and instances.',
```
Add that if between same instance, they just copying links.

percyliang commented:
Code LGTM.

Something still bothers me about the interface (both naming and usage) though.  This is what we (would) have:
- cl wcp worksheet1 worksheet2: appends items from worksheet1 to worksheet2 (copying bundles across instances if necessary)
- cl cp bundle1 worksheet1: copies bundle to worksheet (copying across instances if necessary) - common case
- cl add worksheet1: adds text to current worksheet (doesn't go across instances)
- cl add -m text1: adds text to current worksheet

I think somehow this could be cleaner, given that we're willing to change the interface.  I think the followin would be more logical, although slightly more verbose (avoiding 'copying' is good I think, because most of the time, nothing actually gets copied):
- cl add bundle bundle1
- cl add worksheet worksheet1
- cl add text text1
- cl wadd worksheet1 worksheet2

kashizui commented:
If we would do
`cl add worksheet worksheet1`
then what would be the function of wadd in this case (maybe you meant wcp)?

I like the explicit specification of item type, but your original problem
with `cl add bundle ^ -w worksheet` would still exist. Perhaps this is an
issue of documentation

kashizui commented:
Or maybe we can require specification of the destination worksheet for `cl add` instead of a `-w` flag at all: in what case would we need an "operate completely on this worksheet" flag for `cl add`? We will typically be adding bundles from other worksheets, and this logic doesn't even make sense for adding subworksheets.

`cl add item_type item [item ...] dest_worksheet`

kashizui committed:
Add bundles through cl add again

 - New syntax for cl add: `cl add item_type item [item ...]
   dest_worksheet`
 - Update test-cli
 - Include aliases in `cl help` output

kashizui committed:
Clarify cl cp behavior

kashizui commented:
- New syntax for cl add: `cl add item_type item [item ...]
  dest_worksheet`
- Update test-cli
- Include aliases in `cl help` output
- Clarify `cl cp` help

```
usage: cl add [-h] item_type item_spec [item_spec ...] dest_worksheet

Append text items, bundles, or subworksheets to a worksheet.

positional arguments:
  item_type       type of item(s) to add {text, bundle, worksheet}
  item_spec       item specifications, with the format depending on the specified item_type.
                      text:      (<text>|%<directive>)
                      bundle:    (<uuid>|<name>|^<index>)
                      worksheet: [<alias>::|<address>::]|(<uuid>|<name>)
  dest_worksheet  worksheet to which to add items [<alias>::|<address>::]|(<uuid>|<name>)

optional arguments:
  -h, --help      show this help message and exit
```

kashizui committed:
Clarify `cl add` help

percyliang commented:
I like putting dest_worksheet explicitly.

wadd is different because it adds the items in a worksheet to the
dest_worksheet.

On Fri, Nov 6, 2015 at 11:22 AM, Stephen Koo notifications@github.com
wrote:

> Or maybe we can require specification of the destination worksheet for cl
> add instead of a -w flag at all: in what case would we need an "operate
> completely on this worksheet" flag for cl add? We will typically be
> adding bundles from other worksheets, and this logic doesn't even make
> sense for adding subworksheets.
> 
> `cl add item_type item [item ...] dest_worksheet
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/205#issuecomment-154507933.

kashizui commented:
@percyliang I believe you're referring to `wcp`?

kashizui commented on codalab/lib/bundle_cli.py:
```diff
@@ -502,6 +544,10 @@ def do_command(self, argv):
         argv = self.collapse_bare_command(argv)
         args = parser.parse_args(argv)
 
+        # print args.help
```
remove

percyliang replied:
I propose renaming wcp to wadd.

On Fri, Nov 6, 2015 at 5:54 PM, Stephen Koo notifications@github.com
wrote:

> In codalab/lib/bundle_cli.py
> https://github.com/codalab/codalab-cli/pull/205#discussion_r44205636:
> 
> > @@ -502,6 +544,10 @@ def do_command(self, argv):
> >          argv = self.collapse_bare_command(argv)
> >          args = parser.parse_args(argv)
> > -        # print args.help
> 
> remove
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/205/files#r44205636.

kashizui replied:
Ah ok, makes sense to me.
On Fri, Nov 6, 2015 at 6:01 PM Percy Liang notifications@github.com wrote:

> In codalab/lib/bundle_cli.py
> https://github.com/codalab/codalab-cli/pull/205#discussion_r44205747:
> 
> > @@ -502,6 +544,10 @@ def do_command(self, argv):
> >          argv = self.collapse_bare_command(argv)
> >          args = parser.parse_args(argv)
> > -        # print args.help
> 
> I propose renaming wcp to wadd.
> … <#msg-f:1517144412292410190_>
> On Fri, Nov 6, 2015 at 5:54 PM, Stephen Koo notifications@github.com
> wrote: In codalab/lib/bundle_cli.py <#205 (comment)
> https://github.com/codalab/codalab-cli/pull/205#discussion_r44205636>:
> 
> > @@ -502,6 +544,10 @@ def do_command(self, argv): > argv =
> > self.collapse_bare_command(argv) > args = parser.parse_args(argv) > > + #
> > print args.help remove — Reply to this email directly or view it on GitHub <
> > https://github.com/codalab/codalab-cli/pull/205/files#r44205636>.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/205/files#r44205747.

kashizui committed:
Remove extraneous comments

kashizui commented:
OK Next steps: remove `wcp` and `cp`

kashizui committed:
Remove wcp and cp

kashizui committed:
Clarify help for add and wadd

kashizui commented:
New Interface

```
usage: cl add [-h] [-d] item_type item_spec [item_spec ...] dest_worksheet

Append text items, bundles, or subworksheets to a worksheet. Bundles that do not yet exist on the destination service will be copied over.

positional arguments:
  item_type             type of item(s) to add {text, bundle, worksheet}
  item_spec             item specifications, with the format depending on the specified item_type.
                            text:      (<text>|%<directive>)
                            bundle:    ((<uuid>|<name>|^<index>)|(<alias>|<address>)::(<uuid>|<name>))
                            worksheet: [<alias>::|<address>::](<uuid>|<name>)
  dest_worksheet        worksheet to which to add items [<alias>::|<address>::](<uuid>|<name>)

optional arguments:
  -h, --help            show this help message and exit
  -d, --copy-dependencies
                        if adding bundles, will also add dependencies of the bundles if set
```

```
usage: cl wadd [-h] [-c] source_worksheet_spec dest_worksheet_spec

Append all the items on one worksheet to another. Bundles that do not yet exist on the destination service will be copied over.

positional arguments:
  source_worksheet_spec
                        [<alias>::|<address>::]|(<uuid>|<name>)
  dest_worksheet_spec   [<alias>::|<address>::]|(<uuid>|<name>)

optional arguments:
  -h, --help            show this help message and exit
  -c, --clobber         clobber everything on destination worksheet with source (instead of appending)
```

@percyliang ready for review

kashizui committed:
Fix help doc for add

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -1699,15 +1702,15 @@ def do_wrm_command(self, args):
         arguments=(
             Commands.Argument('source_worksheet_spec', help=WORKSHEET_SPEC_FORMAT, completer=WorksheetsCompleter),
             Commands.Argument('dest_worksheet_spec', help=WORKSHEET_SPEC_FORMAT, completer=WorksheetsCompleter),
             Commands.Argument('-c', '--clobber', help='clobber everything on destination worksheet with source (instead of appending)', action='store_true'),
```
Explicitly say not deleting the bundles on dest worksheet.
Maybe a better name is '--replace'?

percyliang commented:
Nice - LGTM!  Remember to update the docs on the Wiki.

The only lingering hesitation that I have is that 'cl add worksheet foo .' is a bit verbose, but I think it's clear and I'll just get used to it.

kashizui commented:
new `wadd` help:

```
usage: cl wadd [-h] [-r] source_worksheet_spec dest_worksheet_spec

Append all the items of the source worksheet to the destination worksheet, without modifying the existing items on the destination worksheet. Bundles that do not yet exist on the destination service will be copied over.

positional arguments:
  source_worksheet_spec
                        [<alias>::|<address>::](<uuid>|<name>)
  dest_worksheet_spec   [<alias>::|<address>::](<uuid>|<name>)

optional arguments:
  -h, --help            show this help message and exit
  -r, --replace         replace everything on the destination worksheet with the items from the source worksheet, instead of appending
```

kashizui committed:
Rename wadd --clobber to --replace

kashizui commented:
If we allow aliases for the item_type in the future, then people with strong CodaLab-fu could type short commands like `cl a w ^3 0x85817b`

kashizui commented:
Updated wiki:
https://github.com/codalab/codalab/wiki/User_CodaLab-Worksheets-Tutorial

@percyliang could you update the executable papers tutorial (I think it has a mention of `wcp` at the bottom):
https://www.codalab.org/worksheets/0x9b6bb6fdd4aa4bf1b120300fd95f944a/

percyliang commented:
Done.

On Sat, Nov 7, 2015 at 11:38 AM, Stephen Koo notifications@github.com
wrote:

> Updated wiki:
> https://github.com/codalab/codalab/wiki/User_CodaLab-Worksheets-Tutorial
> 
> @percyliang https://github.com/percyliang could you update the
> executable papers tutorial (I think it has a mention of wcp at the
> bottom):
> https://www.codalab.org/worksheets/0x9b6bb6fdd4aa4bf1b120300fd95f944a/
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/205#issuecomment-154741966.
